torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,707 @@
+    void track_module(PyFrameObject* frame);
+
+    // It is imperitive that we do not store strings for each python function,
+    // as that would do terrible things to our profiling overhead. So instead
+    // we store the much cheaper pair of `PyCodeObject*` and `int` which we can
+    // pack into `RawEvent`, and then store a mapping to the full strings the
+    // first time we see a function.
+    //
+    // TODO:
+    //   In theory we should be able to use a combination of Py_INCREF on
+    //   `f_code` and string interning to skip this step. (Effectively reusing
+    //   work that the CPython interpreter has already done.) However it tends
+    //   to segfault and simply caching the strings is inexpensive.
+    struct CodeDescription {
+        CodeDescription(int line_no, std::string filename, std::string funcname)
+            : line_no_(line_no),
+              filename_(std::move(filename)),
+              funcname_(std::move(funcname)) {}
+        int line_no_;
+        std::string filename_;
+        std::string funcname_;
+    };
+
+    struct ModuleForward {
+        ModuleForward(size_t event_index, PyObject* self)
+            : event_index_(event_index), self_(self) {}
+        size_t event_index_;
+
+        // NB:
+        //  This is a non-owning reference to keep `ModuleForward` POD;
+        //  `PythonTracer` owns the contents instead. We  Py_INCREF in
+        //  `track_module`, and `reset` is responsible for  calling Py_DECREF
+        //  when clearing `module_calls_`.
+        PyObject* self_;
+    };
+
+    bool active_;
+    PyObject* module_call_code_;
+    std::string path_prefixes_;
+    std::vector<TraceContext*> trace_contexts_;
+
+    std::vector<RawEvent> events_;
+    std::vector<ModuleForward> module_calls_;
+
+    using DescriptionKey = std::pair</*f_code=*/PyCodeObject*, /*f_lasti=*/int>;
+    ska::flat_hash_map<DescriptionKey, CodeDescription, hash_pair> code_descriptions_;
+};
+
+PythonTracer& PythonTracer::singleton() {
+    static PythonTracer singleton_;
+    return singleton_;
+}
+
+PythonTracer::PythonTracer() : active_(false) {
+    path_prefixes_ = py::module::import("torch.profiler.python_tracer")
+        .attr("_prefix_regex")().cast<std::string>();
+
+    module_call_code_ = py::module::import("torch.nn")
+        .attr("Module")
+        .attr("__call__")
+        .attr("__code__")
+        .ptr();
+}
+
+void PythonTracer::start(size_t max_threads) {
+    TORCH_CHECK(!active_, "PythonTracer is already active")
+    TORCH_CHECK(!trace_contexts_.size(), "PythonTracer should not have active contexts");
+    TORCH_CHECK(max_threads > 0, "max_threads must be positive, got ", max_threads);
+    TORCH_CHECK(
+        max_threads <= max_py_threads,
+        "max_threads must be less than or equal to ", max_py_threads);
+
+    pybind11::gil_scoped_acquire gil;
+    auto t0 = now();
+
+    // Loop over all interpreters and all threads within each interpreter.
```

Comment:
how can you have multiple interpreters? when one forks?

Reply:
It's legal to launch a sub-interpreter. I don't know the exact details of all the ways it can come up. (Python C API docs are a little sparse / not very discoverable in places.)

Reply:
I don't think we support sub-interpreters in pytorch in general.
But torch deploy might manage to embed multiple versions of the python .so in a single process. Is that supported here?

Reply:
I wound up just pulling out the multiple interpreter stuff. If we need it it's not hard to add later.